File: algorithms/[705]设计哈希集合/solution.ts
// topics = ["哈希表"]

// 链地址法
class MyHashSet {
  base: number;
  data: number[][];
  constructor() {
    // 哈希表的大小，素数可以减少规律数据源的影响
    this.base = 769;
    // 数组的每个位置是一个链表，相同哈希值的元素都放入这一链表当中
    this.data = new Array(this.base).fill(0).map(() => new Array());
  }

  // 哈希函数,将集合中任意可能的元素映射到一个固定范围的整数值
  hash(key: number): number {
    return key % this.base;
  }

  add(key: number): void {
    const h = this.hash(key);
    for (const element of this.data[h]) {
      if (element === key) {
        return;
      }
    }
    // 冲突处理
    this.data[h].push(key);
  }

  remove(key: number): void {
    const h = this.hash(key);
    const it = this.data[h];
    for (const [idx, val] of it.entries()) {
      if (val === key) {
        it.splice(idx, 1);
        return;
      }
    }
  }

  contains(key: number): boolean {
    const h = this.hash(key);
    for (const element of this.data[h]) {
      if (element === key) {
        return true;
      }
    }
    return false;
  }
}

/**
 * Your MyHashSet object will be instantiated and called as such:
 * var obj = new MyHashSet()
 * obj.add(key)
 * obj.remove(key)
 * var param_3 = obj.contains(key)
 */

File: algorithms/[705]设计哈希集合/solution.py
# topics = ["设计", "哈希表"]

from typing import List


class MyHashSet:
    """
    链地址法
    """

    def __init__(self):
        # 哈希表的大小，素数可以减少规律数据源的影响
        self.base = 769
        # 数组的每个位置是一个链表，相同哈希值的元素都放入这一链表当中
        self.data: List[List[int]] = [[] for _ in range(self.base)]

    # 哈希函数,将集合中任意可能的元素映射到一个固定范围的整数值
    def hash(self, key: int) -> int:
        return key % self.base

    def add(self, key: int) -> None:
        h = self.hash(key)
        if key in self.data[h]:
            return
        # 冲突处理
        self.data[h].append(key)

    def remove(self, key: int) -> None:
        h = self.hash(key)
        it = self.data[h]
        if key in it:
            it.remove(key)

    def contains(self, key: int) -> bool:
        h = self.hash(key)
        return key in self.data[h]


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)
